kubernetes/nodecollection.go

```diff
@@ -348,34 +348,59 @@ func ensureNodeSource(ctx context.Context, config KubeAgentConfig) (KubeAgentCon
 
 	firstNode := &nodes[0]
 
-	ip, port, err := clientSetNodeSource.NodeAddress(firstNode)
-	if err != nil {
-		return config, fmt.Errorf("error retrieving node addresses: %s", err)
-	}
-
-	if allowDirectConnect(config, nodes) {
-		// test node direct connectivity
-		d := directNodeEndpoints(ip, port)
-		success, err := checkEndpointConnections(config, &nodeHTTPClient, Direct, d.statsSummary())
+	directNodes := 0
+	proxyNodes := 0
+	failedDirect := 0
+	failedProxy := 0
+	directAllowed := allowDirectConnect(config, nodes)
+	for i := range nodes {
+		directlyConnected := false
+		ip, port, err := clientSetNodeSource.NodeAddress(&nodes[i])
 		if err != nil {
-			return config, err
+			return config, fmt.Errorf("error retrieving node addresses: %s", err)
 		}
-		if success {
-			return config, nil
+		if directAllowed {
+			// test node direct connectivity
+			d := directNodeEndpoints(ip, port)
+			success, err := checkEndpointConnections(config, &nodeHTTPClient, Direct, d.statsSummary())
+			if err != nil {
+				log.Warnf("Failed to connect to node [%s] directly with cause [%s]",
+					d.statsSummary(), err.Error())
+				failedDirect++
+			}
+			if success {
+				directlyConnected = true
+				directNodes++
+			}
+		}
+		if !directlyConnected {
+			p := setupProxyAPI(config.ClusterHostURL, firstNode.Name)
+			success, err := checkEndpointConnections(config, &config.HTTPClient, Proxy, p.statsSummary())
+			if err != nil {
+				log.Warnf("Failed to connect to node [%s] via proxy with cause [%s]",
+					p.statsSummary(), err.Error())
+				failedProxy++
+			}
+			if success {
+				proxyNodes++
+			}
 		}
 	}
 
-	// test node connectivity via kube-proxy
-	p := setupProxyAPI(config.ClusterHostURL, firstNode.Name)
-	success, err := checkEndpointConnections(config, &config.HTTPClient, Proxy, p.statsSummary())
-	if err != nil {
-		return config, err
+	log.Infof("Of %d nodes, %d connected directly, %d connected via proxy, and %d could not be reached",
+		len(nodes), directNodes, proxyNodes, failedProxy)
```

Comment:
I like logging the number of nodes that can't be reached. Don't think we need to include directNodes & proxyNodes since it should be an either/or type of situation

Reply:
My thought was just to make it explicitly obvious how we were/were not able to connect, otherwise you have to look at the per-node logs above.